scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/BigQueryIO.scala

```diff
@@ -495,6 +500,103 @@ object TableRowJsonIO {
+) extends ScioIO[T] {
+
+  override val tapT: TapT.Aux[T, Nothing] = EmptyTapOf[T]
+  override type ReadP = Nothing // WriteOnly
+  override type WriteP = BigQueryDynamicTable.WriteParam
+
+  override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+    throw new UnsupportedOperationException("BigQueryDynamicTable is read-only")
+
+  override protected def write(self: SCollection[T], params: WriteP): Tap[Nothing] = {
+    if (self.context.isTest) {
+      throw new NotImplementedError(
+        "BigQuery with dynamic table destinations cannot be used in a test context"
+      )
+    } else {
+      val destinations = DynamicDestinationsUtil.tableFn(tableFn, params.schema)
+
+      var transform = writer.to(destinations)
+
+      if (params.createDisposition != null) {
+        transform = transform.withCreateDisposition(params.createDisposition)
+      }
+      if (params.writeDisposition != null) {
+        transform = transform.withWriteDisposition(params.writeDisposition)
+      }
+
+      transform = params.extendedErrorInfo match {
+        case Disabled => transform
+        case Enabled  => transform.withExtendedErrorInfo()
+      }
+
+      val wr = self.applyInternal(transform)
+
+      params.insertErrorTransform(params.extendedErrorInfo.coll(self.context, wr))
+    }
+    EmptyTap
+  }
+  override def tap(params: Nothing): Tap[Nothing] = EmptyTap
+}
+
+object BigQueryDynamicTable {
+  trait WriteParam {
+    val schema: TableSchema
+    val writeDisposition: WriteDisposition
+    val createDisposition: CreateDisposition
+    val extendedErrorInfo: ExtendedErrorInfo
+    val insertErrorTransform: SCollection[extendedErrorInfo.Info] => Unit
+  }
+
+  object WriteParam extends Writes.WriteParamDefauls {
+    @inline final def apply(
+      schema: TableSchema,
+      wd: WriteDisposition,
+      cd: CreateDisposition,
+      ei: ExtendedErrorInfo
+    )(it: SCollection[ei.Info] => Unit): WriteParam = new WriteParam {
+      val schema: TableSchema = schema
+      val writeDisposition: WriteDisposition = wd
+      val createDisposition: CreateDisposition = cd
+      val extendedErrorInfo: ei.type = ei
+      val insertErrorTransform: SCollection[extendedErrorInfo.Info] => Unit = it
+    }
+
+    @inline final def apply(
+      schema: TableSchema,
+      wd: WriteDisposition = DefaultWriteDisposition,
+      cd: CreateDisposition = DefaultCreateDisposition
+    ): WriteParam = apply(schema, wd, cd, DefaultExtendedErrorInfo)(defaultInsertErrorTransform)
+  }
+
+  def apply[T <: HasAnnotation: Coder](
+    writerFn: T => TableRow,
+    tableFn: ValueInSingleWindow[T] => TableDestination
+  ): BigQueryDynamicTable[T] = {
+
+    val wFn = ClosureCleaner.clean(writerFn)
+    val writer = beam.BigQueryIO
+      .write[T]()
+      .withFormatFunction(Functions.serializableFn(wFn))
+
+    BigQueryDynamicTable(writer)(tableFn)
+  }
+
+  def apply[T <: TableRow: Coder](
+    schema: TableSchema,
```

Comment:
`schema` is not being used and it will be needed down the line to build the `DynamicDestinations`